[RNMobile] Add "Set as Featured Image" Button to Image Block (Android Only)

SiobhyB opened the pull request:
Edit: This PR was closed as it was separated into the following individual PRs:

* **Add Featured Banner to Image Block (Android Only):** #30806
* **Add "Set as Featured Image" Button to Image Block (Android Only):** #31705

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/native-mobile.md -->

SiobhyB committed:
Add "set as featured image" button to image block

This commit adds a "set as featured image" button to the image block's settings and also styles it.

SiobhyB committed:
Add "setFeaturedImage" function in bridge code

With this commit, a "setFeaturedImage" function has been added to the React Native bridge code.

jd-alexander commented on packages/react-native-bridge/android/src/main/java/org/wordpress/mobile/WPAndroidGlue/WPAndroidGlueCode.java:
```diff
@@ -316,6 +316,11 @@ public void requestImageUploadCancel(int mediaId) {
                 mOnMediaLibraryButtonListener.onCancelUploadForMediaDueToDeletedBlock(mediaId);
             }
 
+            @Override
```
Hi @SiobhyB we could also add this override to the [MainApplication](https://github.com/WordPress/gutenberg/blob/702c76e9e9b0de4f96b324f017716afa333f421a/packages/react-native-editor/android/app/src/main/java/com/gutenberg/MainApplication.java) class of the demo project if it's missing. Let me know if you get the project to build in Android Studio. :)

SiobhyB replied:
Added the missing function to MainApplication.java. :)

jd-alexander commented:
@SiobhyB the bridge code is looking good so far 😄 . I am seeing some build errors from the Android side of things and it seems related to the click listeners, so let me know if you are able to resolve them. Thanks 👍

SiobhyB committed:
Add setFeaturedImage function to MainApplication

This commit adds the setFeaturedImage function to the MainApplication file, so that it works as expected with the demo apps.

SiobhyB committed:
Add "OnSetFeaturedImageListener" listener

This commit sets up a listener called "OnSetFeaturedImageListener" in order to allow communication with the native apps.

SiobhyB committed:
Update edit.native.js with small syntax correction

This commit updates edit.native.js with small syntax correction, there was a missing comma at the end of a statement.

SiobhyB committed:
Set up "OnSetFeaturedImageListener" in WPAndroidGlue.java

This commit updates WPAndroidGlue.java in order to set up a "OnSetFeaturedImageListener" function that will communicate with the Android app when a user clicks a "set as featured image" button.

paaljoachim commented:
Hi

Could you add a screenshot?
I wonder if the method could also be transferred over to the desktop Image block.
Thanks.

SiobhyB commented:
👋 @paaljoachim, sure, the mockup of what this will likely look like is as follows: 

![58123648-ef9dfa80-7bd1-11e9-9095-3767b85f69e4 (1)](https://user-images.githubusercontent.com/2998162/107786919-59ffb280-6d46-11eb-90bd-4ccfe124172a.jpeg)

I'll defer to @iamthomasbishop for any extra details around the design and some extra discussion can be found in [the mobile repo](https://github.com/wordpress-mobile/gutenberg-mobile/issues/1011). Hope that helps!

SiobhyB commented:
@paaljoachim, hi! :) I just wanted to follow up again to check whether there are any upcoming plans to implement this on the web? We don't currently have an ETA on the mobile side but would be responsive and adjust our plans to work with the web, if there are firm plans in place there.

paaljoachim commented:
Hi @SiobhyB 

This is the issue
**Add option to set as feature image to image / cover block**
https://github.com/WordPress/gutenberg/issues/13795

I have earlier added a checkbox below the [Featured image placeholder box mockup](https://github.com/WordPress/gutenberg/issues/13795#issuecomment-636787215). Some what similar to the first proposal mockup you share. I did not extend it beyond that point. What about creating another proposal set with focus on web and share them in the above issue? We can then ping @shaunandrews Shaun to give some feedback. This way we will have a have a similar method to automatically add a featured image for web and Mobile views.
Thanks!

SiobhyB committed:
Passing featured image ID from native (Android) to Gutenberg

This commit seeks to send information about a post's featured image ID to Gutenberg by utlising functions from Android.

SiobhyB committed:
Merge branch 'trunk' into update/featured-image-button-in-image-block-settings

SiobhyB committed:
Restore missing curly brace

Accidentally removed a curly brace with the last commit, so restoring it with this one.

SiobhyB committed:
Updates to featured-banner.native.js

This commit including updates to featured-banner.native.js, a component for the "featured" banner that will overlay the image block when an image is featured. "payload.featuredImageId" logs the ID of a featured image after it's updated.

SiobhyB committed:
Attempts to select featured image using getEditedPostAttribute()

This commit includes attempts to select a post's featured image using getEditedPostAttribute().

SiobhyB committed:
Update featured-banner.native.js

With this commit, I'm updating featured-banner.native.js with the latest changes to that component's UI.

SiobhyB committed:
Add featured banner to image/index.native.js

With this commit, a featured banner is added to the image component.

SiobhyB committed:
Update featured image banner and button styles

This commit updates the styles for the featured image banner and button that overlay the image compoonent.

SiobhyB committed:
Create getFeaturedImageId function and cleanup

This commit includes a new getFeaturedImageId function, for grabbing a post's featured image when a component is mounted and also cleans up some unused code.

SiobhyB committed:
Tidy up edit.native.js

hypest commented on packages/react-native-bridge/android/src/main/java/org/wordpress/mobile/WPAndroidGlue/WPAndroidGlueCode.java:
```diff
@@ -559,14 +580,14 @@ public void attachToContainer(ViewGroup viewGroup,
                                   OnGutenbergDidSendButtonPressedActionListener onGutenbergDidSendButtonPressedActionListener,
                                   ShowSuggestionsUtil showSuggestionsUtil,
                                   OnMediaFilesCollectionBasedBlockEditorListener onMediaFilesCollectionBasedBlockEditorListener,
-                                  OnFocalPointPickerTooltipShownEventListener onFocalPointPickerTooltipListener,
```
👋 @SiobhyB , not sure, is the removal of the `onFocalPointPickerTooltipListener` parameter intentional? It seems to introduce a compile error since `onFocalPointPickerTooltipListener` is then not found in line 601. 🤔

SiobhyB replied:
@hypest, I _believe_ that came about by mistake following a previous merge conflict. I had fixed the issue in my local environment but hadn't committed, done that now! Thanks for the ping. :)

SiobhyB committed:
Update AndroidGlueCode

Update AndroidGlueCode to include OnFocalPointPickerTooltipShownEventListener.

SiobhyB committed:
retrigger checks

SiobhyB committed:
Update setup.js

SiobhyB committed:
retrigger checks

hypest commented on packages/react-native-bridge/android/src/main/java/org/wordpress/mobile/WPAndroidGlue/DeferredEventEmitter.java:
```diff
@@ -40,6 +43,8 @@
     private static final String EVENT_NAME_MEDIA_SAVE = "mediaSave";
     private static final String EVENT_NAME_MEDIA_REPLACE_BLOCK = "replaceBlock";
 
+    private static final String EVENT_FEATURED_IMAGE_ID = "featuredImageIdChange";
```
A bit of a nit but, looks like this is a "change happened" event (judging from the "Change" part of the string literal), but the `EVENT_FEATURED_IMAGE_ID` name is a bit more vague. What do you think about renaming to `EVENT_FEATURED_IMAGE_ID_CHANGED` to be more precise about what this event is denoting?

hypest replied:
I think I have a new understanding of what this event is about, please see my newer proposal for the event name in [the comment bellow](https://github.com/WordPress/gutenberg/pull/28854/files#r598733588), sorry for the noise!

SiobhyB replied:
Great, I've responded to the other comment. :)

hypest commented on packages/react-native-bridge/android/src/main/java/org/wordpress/mobile/WPAndroidGlue/DeferredEventEmitter.java:
```diff
@@ -205,6 +210,12 @@ public void onMediaCollectionSaveResult(String firstMediaIdInCollection, boolean
         }
     }
 
+    public void featuredImageIdNotifier(int mediaId) {
```
Similar to my naming comment above, what do you think about renaming to `featuredImageIdChanged()` or better:`onFeaturedImageIdChange` to make it read like an event handler? The original `featuredImageIdNotifier` makes me wonder what gets notified... like, is it the user that receives a notification when this function is called?

hypest replied:
Reading more code where this function is used, I see that it's also called when the JS side [just asks for an update of the featured image id](https://github.com/wordpress-mobile/WordPress-Android/blob/2700d6c49232e6866abdc796cf7cefb5f8c26dea/libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/gutenberg/GutenbergEditorFragment.java#L296). So, the function's purpose is not only to notify the JS side that the id has changed, which means that the event name ([defined here](https://github.com/WordPress/gutenberg/pull/28854/files#diff-bd64b4a30ececb510793aeabfe0b68175eaa825a9f121d8dd953e688a355c01fR46)) is not actually accurate 🤔.

So, if I understand things properly, the event is all in all just trying to pass the the media id of the current featured image, right, whether it was recently updated or not? If so, what do you think about naming the event with `EVENT_FEATURED_IMAGE_ID_CURRENT = "featuredImageIdCurrent"`?

SiobhyB replied:
> The original featuredImageIdNotifier makes me wonder what gets notified... like, is it the user that receives a notification when this function is called?

This is useful feedback, thank you @hypest! I looked through other similar functions and renamed `featuredImageIdNotifier` to `onRequestFeaturedImageId` in order to hopefully be clearer of the function's purpose. Let me know if you think this is still unclear!

> So, if I understand things properly, the event is all in all just trying to pass the the media id of the current featured image, right, whether it was recently updated or not? If so, what do you think about naming the event with EVENT_FEATURED_IMAGE_ID_CURRENT = "featuredImageIdCurrent"?

That makes sense! The purpose changed as the code evolved (it was originally only passing the id when the image changed) which is why I think the naming became a bit messy/confusing. This is a good lesson to keep on top and iterating on names. Thank you! I've updated this to "featuredImageIdCurrent" now.

hypest replied:
> I looked through other similar functions and renamed featuredImageIdNotifier to onRequestFeaturedImageId in order to hopefully be clearer of the function's purpose. Let me know if you think this is still unclear!

Aha.

Still trying to grok what the code is trying to do, please bear with me :). There are two rather different code paths that use this function:
1. The JS side asks the native side to provide the featured image id ([getter named function here](https://github.com/WordPress/gutenberg/blob/46ba7e8d770cc0cc1a93664047fdcaaa12f0f0f0/packages/react-native-bridge/android/src/main/java/org/wordpress/mobile/ReactNativeGutenbergBridge/RNReactNativeGutenbergBridgeModule.java#L237-L239)). The native side then [calls `onRequestFeaturedImageId` only if the media id provided matches the featured image id on file](https://github.com/wordpress-mobile/WordPress-Android/blob/622dfad0700d288349fd344783b3bbf24e242cdc/libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/gutenberg/GutenbergEditorFragment.java#L349-L353). By the way, looks like it's not a getter after all since the response to the original `getFeaturedImageId` is conditional. Looks to me more like a "check for something", not a plain getter. If so, would it make sense to rename getFeaturedImageId -> checkIfFeaturedImage?
2. The native side [sends the id when the featured image is set](https://github.com/wordpress-mobile/WordPress-Android/blob/622dfad0700d288349fd344783b3bbf24e242cdc/WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java#L2554).

From flow No2 above, and the function implementation itself ([native side](https://github.com/WordPress/gutenberg/blob/d8c0824c087e5fc76e955123665bc9f2afc893e8/packages/react-native-bridge/android/src/main/java/org/wordpress/mobile/WPAndroidGlue/DeferredEventEmitter.java#L214-L216), https://github.com/WordPress/gutenberg/blob/75de59fd6cacd8faeb787e20c8bc218c24992f46/packages/block-library/src/image/edit.native.js#L305-L319) I understand that onRequestFeaturedImageId just sends the current featured image id to the JS side so, may I propose renaming it to `sendToJSFeaturedImageId`?

Then, flow No1 shows to me we need to better name and document why `onGetFeaturedImageId` doesn't always call sendToJSFeaturedImageId.
Does this all make sense? Sorry if I'm a bit confused on this one.

SiobhyB replied:
@hypest, thank you again for your help here!  

> By the way, looks like it's not a getter after all since the response to the original getFeaturedImageId is conditional. Looks to me more like a "check for something", not a plain getter. If so, would it make sense to rename getFeaturedImageId -> checkIfFeaturedImage?
>
> Then, flow No1 shows to me we need to better name and document why onGetFeaturedImageId doesn't always call sendToJSFeaturedImageId.

It's right that the `getFeaturedImageId` function is only checking whether an image is featured, I've changed its name to `checkIfFeaturedImage` for the time being.

To further clarify: The `checkIfFeaturedImage` function is called within `componentWillUnmount()` on the JS side. It's currently responsible for flagging which image is featured when the editor is first loaded.

The background for this is that I was unable to get `getEditedPostAttribute` to work to get the post's initial featured image (discussed on project thread and in project's Slack channel). 

I've spent a bit of time with this but am not sure if there's a more elegant way to achieve the same effect, but it's entirely likely I've been looking at the code too long! I'd be interested to hear whether you think this approach makes sense, with those extra details, or if I should work on this a bit more.

hypest replied:
> To further clarify: The checkIfFeaturedImage function is called within componentWillUnmount() on the JS side. It's currently responsible for flagging which image is featured when the editor is first loaded.

Ah, I see 👍, I will check this out now that I read the logic better. Just to make sure, do you perhaps meant to say `componentDidMount` since that's the one called on editor start?

> I've spent a bit of time with this but am not sure if there's a more elegant way to achieve the same effect, but it's entirely likely I've been looking at the code too long! I'd be interested to hear whether you think this approach makes sense, with those extra details, or if I should work on this a bit more.

Will have a look from than angle and let you know! Can you update the WPAndroid side as well as I think it doesn't compile currently due to bridge changes, thanks! ([Commented on the WPAndroid PR as well](https://github.com/wordpress-mobile/WordPress-Android/pull/14052#issuecomment-810152596))

SiobhyB replied:
> Just to make sure, do you perhaps meant to say componentDidMount since that's the one called on editor start?

Ah, yes, careless copy/paste, sorry! 🤦‍♀️ I meant `componentDidMount`.

> Can you update the WPAndroid side as well as I think it doesn't compile currently due to bridge changes, thanks!

Updated! :)

hypest replied:
> Will have a look from than angle and let you know! Can you update the WPAndroid side as well as I think it doesn't compile currently due to bridge changes, thanks! (Commented on the WPAndroid PR as well)

Just an update. I had a short look and I feel like triggering a bridge method to check each image in the post is too much and it might not scale well in image heavy posts. Feels to me that we should pass the initial value of the featured image id during editor boot, find a way to keep that in the global(?) state and check against that for every image block.

To that end, I'm having another look to the original idea around `getEditedPostAttribute` to see if we could hook at that place, to put the id in the global post object, for example. I'm not familiar with that part of the store though so, will continue exploring into it.

hypest replied:
Did a quick dive and I think that adding a `featured_media: props.featuredMedia` in [the post object we prepare when starting the editor](https://github.com/WordPress/gutenberg/blob/trunk/packages/edit-post/src/editor.native.js#L122-L136), accompanied by passing that `featuredMedia` id [down the bridge](https://github.com/WordPress/gutenberg/blob/trunk/packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/WPAndroidGlue/GutenbergProps.kt), is then making `featured_media` available to the `getEditedPostAttribute` Store function. Not sure if that's something you already had tried recently @SiobhyB , let me know if it helps.

That's something to help with the bootstrapping of the editor with the featured image id becoming available then, but will need some more treatment to enable _updating_ the id when the featured image gets set after the editor has already started.

SiobhyB replied:
@hypest, thank you for this! 🙇‍♀️ I'm working on getting this to work as expected now. I've committed my latest work in progress, which is not yet working, here: https://github.com/WordPress/gutenberg/pull/28854/commits/0cdeadb1ddf8cbc4c491e2eb21b42eda683ac247

I also made the assumption that this will require some changes to the Android side, and have made some initial changes here: https://github.com/wordpress-mobile/WordPress-Android/pull/14052/commits/0ec24df06d5ff419e0f7a21849c1e53afdbf70a8

Does this seem like I'm on the right path? I'm happy to continue troubleshooting and trying to figure this out, but just wanted to make sure I wasn't going off in the wrong direction.

hypest replied:
Oh, glad you liked the proposal and started going down that route!

Looks like a promising path so far so yeah, please feel free to go ahead. A disclaimer here though: I haven't really worked on getting a real feel of how the _updating_ part will work this time so, please keep that in mind in case this new route proves impractical too after all. Thanks!

SiobhyB replied:
I've been able to get this working to grab the post's initial featured image ID now and have made use of `setAttributes` to get this working in ComponentDidUpdate [here](https://github.com/WordPress/gutenberg/pull/28854/files#diff-7126ae03000a022bcefa0d2a5fb03b6aca56ed4047e3a49d2131babf229cf6c1R197). How does this seem? Looking forward to any feedback or ideas on this one. :)

hypest replied:
> How does this seem?

Yeah, I think we're on a good route here. We chatted a bit over Zoom, to share my thoughts on how the ImageEdit component just needs to listen to the store updates and compare the featured image id it gets from there with the id it keeps in the block attributes, and use those to compute a "isFeaturedImage" boolean inside the render() function.

Then, it becomes a matter of figuring out how to also do the update of the id in the store, for which we can use the [editEntityRecord()](https://github.com/WordPress/gutenberg/blob/trunk/packages/core-data/src/actions.js#L248) action creator (acquiring it via the `withDispach`. But, that part (to update the featured image id value when the image changes) can be done in a separate PR I think.

Let me know what you think!

SiobhyB replied:
Thank you again for this! I've spent a bit of time today trying to process this and have a few questions:

* What are the overall benefits to moving the `setFeaturedImageState()` functionality inside the `render()` function vs. the current approach? 
* `setFeaturedImageState()` is currently called in `componentDidMount()` for when the editor first loads and then `componentDidUpdate()` for when an image is replaced directly within the image block (using the Edit icon to the upper right when the block is selected). Will moving the functionality for setting state from `setFeaturedImageState()` to `render()` resolve the need to do that? Is that one of the benefits?
* It seems like I'll need to figure out how to use `editEntityRecord()` to grab the post's current featured image ID _before_ I can move the functionality for setting state from  `setFeaturedImageState()`. Is that right?

Sorry for my fuzziness on this one, just want to make sure I'm clear before moving ahead!

SiobhyB replied:
It seems all I had to do was ask those questions and then play around a little more to figure out the answers. :) I've been able to convert the functionality previously included in `setFeaturedImageState` to an `isFeaturedImage` boolean. I can see now that the answer to my second question is "yes" and the code is a lot cleaner as a result. 🎉 I've committed those changes.

hypest replied:
Nice to see the simplification of `isFeaturedImage`!

> setting an isFeaturedImage state

Just a quick comment on the terminology used in the [commit message](https://github.com/WordPress/gutenberg/pull/28854/commits/39303e4a13dce66bfeed96060e452af72e65fb53): now that `isFeaturedImage` is just a helper local const in `render()`, referring to it as "state" can be misleading, mainly because within React, "state" has a special and specific [meaning](https://reactjs.org/docs/state-and-lifecycle.html). I'd recommend referring to it as "local variable/const" to avoid confusion with React's use.

hypest commented on packages/react-native-bridge/android/src/main/java/org/wordpress/mobile/WPAndroidGlue/WPAndroidGlueCode.java:
```diff
@@ -172,6 +178,10 @@ public boolean isContentChanged() {
         void onQueryCurrentProgressForSavingMedia();
     }
 
+    /* public interface OnSetFeaturedImageListener {
```
Hey Siobhan, aha, can you elaborate on why this new interface is commented out and the new featured image related interface functions are placed in `OnMediaLibraryButtonListener`? I'm not sure if `OnMediaLibraryButtonListener` is actually related 🤔, I'd expect this new/separate interface definition to be the desired implementation.

SiobhyB replied:
@hypest, this one was under the "I'll get to that later!" list in my mind that I hadn't got round to yet. I believe I was just exploring a few things to see if they worked when I placed the functions under `OnMediaLibraryButtonListener` and hadn't got round to implementing `OnSetFeaturedImageListener` yet. I've done that now. :) Thanks for the push!

SiobhyB committed:
Update function name from "featuredImageIdChange" to "featuredImageIdCurrent"

With this commit, I'm updating a function's name from "featuredImageIdChange" to "featuredImageIdCurrent" in order to more accurately reflect its purpose.

SiobhyB committed:
Rename "featuredImageIdNotifier" to "onRequestFeaturedImageId"

This commit renames "featuredImageIdNotifier" to "onRequestFeaturedImageId" in order to clarify the function's purpose.

SiobhyB committed:
Create OnSetFeaturedImageListener

This commit introduces an "OnSetFeaturedImageListener" and moves new featured image related functions to it from "OnMediaLibraryButtonListener".

SiobhyB committed:
Dismiss bottomsheet when "setFeaturedImage" is called

SiobhyB committed:
Introduce "onRemoveFeatured" button and tidying up names

This commit introduces a "onRemoveFeatured" button that displays in the image block's settings when an image is featured. It also changes the name of the "onGetFeaturedImageId" function to "checkIfFeaturedImage", to better reflect its current functionality.

SiobhyB committed:
Updates to function names

This commit updates "onRequestFeaturedImageId" to "sendToJSFeaturedImageId" and "getFeaturedImageId" to "checkIfFeaturedImage". In both cases, this has been done to clarify the purpose of the functions.

SiobhyB committed:
Convert FeaturedBanner to Badge component

This commit converts the FeaturedBanner component to a more generic Badge component. This is to make it more re-usable. There are also some small styling tweaks included in this commit, including changes to the component's border-radius.

SiobhyB committed:
Update edit.native.js to correct merge conflict

SiobhyB committed:
Correct merge conflict with GutenbergBridgeJS2Parent.java

This commit is an attempt to correct this PR's merge conflict with GutenbergBridgeJS2Parent.java.

SiobhyB committed:
retrigger checks

SiobhyB committed:
retrigger checks

SiobhyB committed:
retrigger checks

SiobhyB committed:
retrigger checks

SiobhyB committed:
Update edit.native.js to fix merge conflict

SiobhyB committed:
Update edit.native.js to check if an image is featured uppn changes

This commit fixes a bug, where the "featured" banner did not update if an image was replaced directly within a block, using the available settings to the upper right.

SiobhyB committed:
Update styles.native.scss to add border to the top of featured button

This commit also tidies up some of the code around styling to account for dark mode.

SiobhyB committed:
Update edit.native.js

This commit moves the "checkIfFeaturedImage" check into its own function.

SiobhyB committed:
Introduce "featuredMedia" prop for use when editor loads

This commit introduces a "featuredMedia" prop that is intended to be used when Gutenberg first loads. The prop is going to be passed over the bridge so that we can get information about a post's featured image ID from the app.

SiobhyB committed:
Fetch ID of post's initial featured image

This commit introduces some changes in react-native-editor/src/index.js that enable getEditedPostAttribute( 'featured_media' ) to fetch the initial ID of a post's featured image. This is then used in the image block's componentDidMount function to pinpoint a featured image when it's first loaded in the editor.

SiobhyB committed:
Merge branch 'trunk' into update/featured-image-button-in-image-block-settings

SiobhyB committed:
Updated featured state when image is replaced within block

This commit uses "setAttributes" to define a "currentFeaturedImageId" attribute. This attribute is then used in "componentDidUpdate" to check if an image is featured if it's replaced directly within the image block. There are also some smaller tweaks/improvements to the code in this commit, such as changing an if/else statement to use ES6's shorthand.

SiobhyB committed:
Remove checkIfFeaturedImage function

This function has been replaced with functionality to grab a post's initial ID, as can be seen in the previous couple of commits.

SiobhyB committed:
Fix merge conflict

SiobhyB committed:
Fix merge conflict

SiobhyB committed:
Add "setFeaturedImage" and "featuredImageIdCurrent" functions to Swift files

SiobhyB committed:
Simplify this.state.isFeaturedImage by converting to a boolean insider render()

This commit simplifies the process of determining whether an image is featured (currently done via by setting an isFeaturedImage state) by converting it to a boolean insider the render() function.

SiobhyB committed:
Comment out iOS-specific changes to the bridge

This commit comments out iOS-specific changes to the bridge in ordet to test whether these changes are the reason behind a failing performance test.

SiobhyB committed:
Add iOS functions to bridge

SiobhyB committed:
Flag featured image when editor mounts and comment out iOS methods from bridge

This commit adds a check to "componentDidMount" so that a featured image is flagged when the editor first mounts (fixes previous regression). It also comments out iOS methods over the bridge, a temporary change.

SiobhyB committed:
Comment out a bridge function for iOS (missing from previous commit).

SiobhyB committed:
Add set featured image functionality behind dev flag

SiobhyB committed:
Move "set as featured" button behind devOnly and androidOnly flag

This commit also changes the "featured" banner so that it's accessible to all.

SiobhyB committed:
Calling "__DEV__" directly instead of defining a separate variable

SiobhyB committed:
Update changelog.md

SiobhyB commented:
@hypest, @iamthomasbishop, the current PR includes the **Featured** banner for Android users. It then also adds the **set as featured** functionality to the image block's settings for Android users in development mode. The screenshots in this issue's main description are up-to-date for Android. If this looks good to merge to you, I'll then work on the following in separate PRs:

* Introduce the **Featured** banner for iOS users.
* Introduce the **set as featured** functionality to the image block's settings for iOS users.
* Improve the way the featured image ID is managed on the JS side.
* Remove the dev-only flag.

Looking forward to hearing what you think!

SiobhyB committed:
Merge branch 'trunk' into update/featured-image-button-in-image-block-settings

SiobhyB commented:
Following a recent discussion, I've now split the **Featured** banner part of this off into its own PR here: https://github.com/WordPress/gutenberg/pull/30806

SiobhyB committed:
Merge branch 'trunk' into update/featured-image-button-in-image-block-settings

SiobhyB committed:
Merge branch 'trunk' into update/featured-image-button-in-image-block-settings

SiobhyB committed:
Remove "subscribeFeaturedImageIdCurrent" functionality

This commit removes "subscribeFeaturedImageIdCurrent", as it's no longer in use.

SiobhyB committed:
Remove featuredImageId check in componentDidMount()

SiobhyB committed:
Remove isFeaturedImage functionality from image/index.native.js

This functionality was moved to image/edit.native.js.

SiobhyB committed:
Remove iOS functions from bridge

SiobhyB committed:
Remove android.util.Log import

SiobhyB committed:
Remove redundant space

SiobhyB committed:
Update CHANGELOG.md

SiobhyB committed:
Tweak androidOnly const

SiobhyB committed:
Merge branch 'trunk' into update/featured-image-button-in-image-block-settings

SiobhyB committed:
Refactor 'set as featured' button into its own function

SiobhyB committed:
Merge branch 'update/featured-image-button-in-image-block-settings' of github.com:SiobhyB/gutenberg into update/featured-image-button-in-image-block-settings

SiobhyB committed:
Merge branch 'trunk' into update/featured-image-button-in-image-block-settings

SiobhyB committed:
Fix syntax issue with onSetFeatured function

SiobhyB committed:
Merge remote-tracking branch 'upstream/trunk' into update/featured-image-button-in-image-block-settings

SiobhyB commented:
@jd-alexander, this chain of PRs is ready for review, whenever you have the chance to take a look. Thank you in advance! 🙇‍♀️

jd-alexander commented:
> @jd-alexander, this chain of PRs is ready for review, whenever you have the chance to take a look. Thank you in advance! 🙇‍♀️

Thanks! I will have a look.

SiobhyB committed:
Merge branch 'WordPress:trunk' into update/featured-image-button-in-image-block-settings

hypest commented on packages/block-library/src/image/edit.native.js:
```diff
@@ -443,13 +502,13 @@ export class ImageEdit extends Component {
 		} = this.props;
 		const { align, url, alt, id, sizeSlug, className } = attributes;
 
+		const isFeaturedImage = featuredImageId === attributes.id;
```
👋 Siobhan! I see the [same line in trunk already](https://github.com/WordPress/gutenberg/blob/f14962f283f63039be1e790f552489a1705d2928/packages/block-library/src/image/edit.native.js#L451) so, I assume the PR needs some updating from trunk? While at it, since by now you have write access to the Gutenberg repo, it makes sense to re-open this PR as one directly on the repo instead of the fork, that will speed up the review process, wdyt?

SiobhyB replied:
Thank you for pointing that out, it looks like the line got moved about at some point in this PR, I'll rework that.

> it makes sense to re-open this PR as one directly on the repo instead of the fork, that will speed up the review process, wdyt?

This makes sense to me and will likely simplify the process of cleaning up the commit history! I'll get about doing that and ping you when the new PR is ready for review. :)

hypest commented on packages/block-library/src/image/edit.native.js:
```diff
@@ -631,5 +691,12 @@ export default compose( [
+	withDispatch( ( dispatch ) => {
+		return {
+			closeSettingsBottomSheet() {
+				dispatch( 'core/edit-post' ).closeGeneralSidebar();
```
Ah, [like recently](https://github.com/WordPress/gutenberg/pull/30806#discussion_r613207635), let's not use store name literals directly, thanks!

SiobhyB replied:
Ah, thank you! I'll refactor this in the new PR I'm creating for the issue. :)

SiobhyB commented:
Closing this PR, as it was separated into the following:

* **Add Featured Banner to Image Block (Android Only):** https://github.com/WordPress/gutenberg/pull/30806
* **Add "Set as Featured Image" Button to Image Block (Android Only):** https://github.com/WordPress/gutenberg/pull/31705
